File: Linked List/Arrays/4. Median of Two Sorted Arrays/fighting.cpp
#include <vector>
#include <iostream>
using namespace std;

class Solution
{
public:
    double findMedianSortedArrays(vector<int> &nums1, vector<int> &nums2)
    {
        int total = nums1.size() + nums2.size();
        if (total % 2 == 1)
        {
            return findKstNumber(nums1, nums2, total / 2 + 1);
        }
        else
        {
            return (findKstNumber(nums1, nums2, total / 2) + findKstNumber(nums1, nums2, total / 2 + 1)) / 2.0;
        }
    }

    double findKstNumber(vector<int> &nums1, vector<int> &nums2, int k)
    {
        if (nums1.size() > nums2.size())
        {
            return findKstNumber(nums2, nums1, k);
        }

        if (nums1.size() == 0)
        {
            return nums2[k - 1];
        }

        if (k == 1)
        {
            return min(nums1[0], nums2[0]);
        }

        int mid1 = min(int(k / 2), int(nums1.size()));
        int mid2 = k - mid1;

        if (nums1[mid1 - 1] > nums2[mid2 - 1])
        {
            vector<int> newNums2 = vector<int>(nums2.begin() + mid2, nums2.end());
            return findKstNumber(nums1, newNums2, k - mid2);
        }
        else if (nums1[mid1 - 1] < nums2[mid2 - 1])
        {
            vector<int> newNums1 = vector<int>(nums1.begin() + mid1, nums1.end());
            return findKstNumber(newNums1, nums2, k - mid1);
        }
        else
        {
            return nums1[mid1 - 1];
        }
    }
};

int main()
{

    int x[] = {1, 2, 3, 4, 5, 6, 7, 8};
    int y[] = {1, 2, 3, 4, 5};
    vector<int> a;
    vector<int> b;
    a.assign(x, x + sizeof(x) / sizeof(int));
    b.assign(y, y + sizeof(y) / sizeof(int));
    Solution temp;
    cout << temp.findMedianSortedArrays(a, b) << endl;
}

File: Linked List/Arrays/4. Median of Two Sorted Arrays/fighting.py
class Solution:
    def findMedianSortedArrays(self, nums1: list[int], nums2: list[int]) -> float:
        total = len(nums1) + len(nums2)
        if total % 2 == 0:
            return (self.findKstNumber(nums1, nums2, total // 2) + self.findKstNumber(nums1, nums2, total // 2 + 1)) / 2
        else:
            return self.findKstNumber(nums1, nums2, total // 2 + 1)

    def findKstNumber(self, nums1: list[int], nums2: list[int], k: int) -> float:
        if len(nums1) > len(nums2):
            return self.findKstNumber(nums2, nums1, k)

        if len(nums1) == 0:
            return nums2[k - 1]

        if k == 1:
            return min(nums1[0], nums2[0])

        mid1 = min(k // 2, len(nums1))
        mid2 = k - mid1

        if nums1[mid1 - 1] < nums2[mid2 - 1]:
            return self.findKstNumber(nums1[mid1:], nums2, k - mid1)
        elif nums1[mid1 - 1] > nums2[mid2 - 1]:
            return self.findKstNumber(nums1, nums2[mid2:], k - mid2)
        else:
            return nums1[mid1 - 1]


nums1 = [1, 2, 3, 4, 5, 6, 7, 8]
nums2 = [1, 2, 3, 4, 5]
temp = Solution()
print(temp.findMedianSortedArrays(nums1, nums2))
